pkg/querier/batch/stream.go

```diff
@@ -18,82 +16,108 @@ import (
+	return len(bs.batches)
 }
 
-func (bs *batchStream) at() (int64, float64) {
-	b := &(*bs)[0]
-	return b.Timestamps[b.Index], b.Values[b.Index]
+func (bs *batchStream) reset() {
+	for i := range bs.batches {
+		bs.batches[i].Index = 0
+	}
 }
 
-func (bs *batchStream) atHistogram() (int64, unsafe.Pointer) {
-	b := &(*bs)[0]
-	return b.Timestamps[b.Index], b.PointerValues[b.Index]
+func (bs *batchStream) hasNext() chunkenc.ValueType {
+	if bs.len() > 0 {
+		return bs.curr().ValueType
+	}
+	return chunkenc.ValNone
+}
+
+func (bs *batchStream) next() {
+	b := bs.curr()
+	b.Index++
+	if b.Index >= b.Length {
+		bs.batches = bs.batches[1:]
+	}
 }
 
-func (bs *batchStream) atFloatHistogram() (int64, unsafe.Pointer) {
-	b := &(*bs)[0]
-	return b.Timestamps[b.Index], b.PointerValues[b.Index]
+func (bs *batchStream) curr() *chunk.Batch {
+	return &bs.batches[0]
 }
 
-// mergeStreams merges streams of Batches of the same series over time.
+// merge merges this streams of chunk.Batch objects and the given chunk.Batch of the same series over time.
 // Samples are simply merged by time when they are the same type (float/histogram/...), with the left stream taking precedence if the timestamps are equal.
 // When sample are different type, batches are not merged. In case of equal timestamps, histograms take precedence since they have more information.
-// hPool and fhPool are pools of pointers where the discarded histogram.Histogram and histogram.FloatHistogram objects from left and right streams
-// will be placed, so that they can be reused. If hPool and fhPool are nil, the discarded histogram.Histogram and histogram.FloatHistogram objects
-// will not be reused.
-func mergeStreams(left, right, result batchStream, size int, hPool *zeropool.Pool[*histogram.Histogram], fhPool *zeropool.Pool[*histogram.FloatHistogram]) batchStream {
-
+func (bs *batchStream) merge(batch chunk.Batch, size int) {
 	// Reset the Index and Length of existing batches.
-	for i := range result {
-		result[i].Index = 0
-		result[i].Length = 0
+	for i := range bs.batchesBuf {
+		bs.batchesBuf[i].Index = 0
+		bs.batchesBuf[i].Length = 0
 	}
 
 	resultLen := 1 // Number of batches in the final result.
-	b := &result[0]
+	b := &bs.batchesBuf[0]
 
 	// Step to the next Batch in the result, create it if it does not exist
 	nextBatch := func(valueType chunkenc.ValueType) {
 		// The Index is the place at which new sample
 		// has to be appended, hence it tells the length.
 		b.Length = b.Index
 		resultLen++
-		if resultLen > len(result) {
+		if resultLen > len(bs.batchesBuf) {
 			// It is possible that result can grow longer
 			// then the one provided.
-			result = append(result, chunk.Batch{})
+			bs.batchesBuf = append(bs.batchesBuf, chunk.Batch{})
 		}
-		b = &result[resultLen-1]
+		b = &bs.batchesBuf[resultLen-1]
 		b.ValueType = valueType
 	}
 
-	populate := func(s batchStream, valueType chunkenc.ValueType) {
+	populate := func(batch chunk.Batch, valueType chunkenc.ValueType) {
```

Comment:
batchStream was slice so the input would have been passed by reference but here we're passing an object that will be copied. I wonder about the performance impact of this extra copy for every sample. Please take a benchmark compare.

I think we should pass `*chunk.Batch` to populate. Would be quite natural for the left side and we can make the code pass pointer for the right side as well.

Reply:
Good catch, thank you.
I have fixed that.